File: Vandame/ChatGPT_MATtoPY/G1sound_mean.py
import numpy as np
from scipy.fft import fft, ifft
from scipy.signal import tukey

sons_moy = [None] * len(sons)
fft_moy = [None] * len(sons)

f = 0
for i in range(len(sons)):
    f += 1
    dim = 0
    for j in range(len(sons[i])):
        if sons[i][j] is not None:
            dim = max(dim, sons[i][j].shape[0])

    for j in range(len(sons[i])):
        if sons[i][j] is not None:
            FFT = np.abs(fft(sons[i][j], dim))
            if sons[i][j].shape[0] == dim:
                phase = np.angle(fft(sons[i][j], dim))
            if fft_moy[i] is None:
                fft_moy[i] = FFT / np.linalg.norm(FFT)
            else:
                fft_moy[i] += FFT / np.linalg.norm(FFT)

    fft_moy[i] /= np.linalg.norm(fft_moy[i])
    sons_moy[i] = np.real(ifft(fft_moy[i] * np.exp(1j * phase)))

    sons_moy[i] *= tukey(len(sons_moy[i]), alpha=0.25)

    if location[i]:
        t = np.arange(Data.wavinfo.SampleRate * 15 + 1)
        t = np.mod(t, round(Data.wavinfo.SampleRate * max(Data.wavinfo.Duration) /
                            Data.spectroinfo.s[1] * 1 / Frequel[location[i][-1]])) == 0

        sons_moy[i] = np.convolve(t.astype(float), sons_moy[i])

# Clear variables
sons_moy = None
fft_moy = None
dim = None
FFT = None
location = None
phase = None
Frequel = None
Fminfiltre = None
Fmaxfiltre = None
t = None
bandmin = None
f = None
fft_moy = None
sons = None

File: Vandame/matlab_code/G1sound_mean.m
sons_moy = cell(size(sons,1),1);
fft_moy = cell(size(sons,1),1);

        f = waitbar(0,'sound mean');
for i = 1:size(sons,1)
    waitbar(i/size(sons,1),f);

    dim = 0;
    for j = 1 : size(sons,2)
        if ~isempty(sons{i,j})
            dim=max(dim,size(sons{i,j},1));
        end
    end
    
    for j = 1 : size(sons,2)
        if ~isempty(sons{i,j})
            FFT = abs(fft(sons{i,j},dim));
            if size(sons{i,j},1) == dim
                phase = angle(fft(sons{i,j},dim));
            end
            if isempty(fft_moy{i})
                fft_moy{i}=FFT/norm(FFT);
            else
                fft_moy{i}=fft_moy{i}+FFT/norm(FFT);
            end
        end
    end
    fft_moy{i} = fft_moy{i}/norm(fft_moy{i});
    sons_moy{i} = real(ifft(fft_moy{i}.*exp(1i*phase)));
    
    sons_moy{i} = sons_moy{i}.*tukeywin((length(sons_moy{i})),0.25);
    %% repetition of the sound
    if ~isempty(location{i})
        t = 0:max(Data.wavinfo.SampleRate*15);
        t = double( mod(t,round(Data.wavinfo.SampleRate*...
            max(Data.wavinfo.Duration)/Data.spectroinfo.s(2)*1/(Frequel(location{i}(end))))) == 0 ).';

        sons_moy{i} = conv(t,sons_moy{i});
        
    end
    
end

close(f);
%%
clear ans dim FFT location i j k phase Frequel Fminfiltre Fmaxfiltre t bandmin f fft_moy 
clear sons